pyphare/pyphare_tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,528 @@
+import unittest
+
+from ddt import ddt, data
+import pyphare.core.box as boxm
+from pyphare.core.box import Box
+from pyphare.pharesee.particles import Particles
+from pyphare.pharesee.hierarchy import FieldData
+from pyphare.pharesee.hierarchy import ParticleData
+from pyphare.pharesee.hierarchy import PatchHierarchy
+from pyphare.pharesee.hierarchy import Patch, PatchLevel
+from pyphare.pharesee.geometry import particle_ghost_area_boxes
+from pyphare.pharesee.geometry import level_ghost_boxes, hierarchy_overlaps, touch_domain_border
+from pyphare.core.gridlayout import GridLayout
+
+import numpy as np
+
+
+def bx(ghost_box, dx, Lx, origin):
+    x = origin + np.arange(ghost_box.size() + 1) * dx - 5 * dx
+    return np.sin(2 * np.pi / Lx * x)
+
+
+def by(ghost_box, dx, Lx, origin):
+    x = origin + np.arange(ghost_box.size()) * dx - 5 * dx
+    return np.cos(2 * np.pi / Lx * x)
+
+
+def bz(ghost_box, dx, Lx, origin):
+    x = origin + np.arange(ghost_box.size()) * dx - 5 * dx
+    return np.sin(4 * np.pi / Lx * x)
+
+
+def ex(ghost_box, dx, Lx, origin):
+    x = origin + np.arange(ghost_box.size()) * dx - 5 * dx
+    return np.sin(2 * np.pi / Lx * x)
+
+
+def ey(ghost_box, dx, Lx, origin):
+    x = origin + np.arange(ghost_box.size() + 1) * dx - 5 * dx
+    return np.cos(2 * np.pi / Lx * x)
+
+
+def ez(ghost_box, dx, Lx, origin):
+    x = origin + np.arange(ghost_box.size() + 1) * dx - 5 * dx
+    return np.sin(4 * np.pi / Lx * x)
+
+
+def build_hierarchy(**kwargs):
+    """accepted keywords:
+     - simulation : a simulation Object
+
+     or
+
+     - nbr_cells
+     - origin
+     - interp_order
+     - domain_size
+     - cell_width
+     - refinement_ratio
+     - refinement_boxes
+     """
+    if "simulation" in kwargs:
+        for k in kwargs:
+            if k != "simulation":
+                print("warning: 'simulation' given, {} discarded".format(k))
+
+        sim = kwargs["simulation"]
+        nbr_cells = sim.cells[0]
+        origin = sim.origin[0]
+        interp_order = sim.interp_order
+        domain_size = sim.simulation_domain()[0]
+        cell_width = sim.dl[0]
+        refinement_ratio = 2
+        refinement_boxes = sim.refinement_boxes
+
+    else:
+        nbr_cells = kwargs["nbr_cells"]
+        origin = kwargs["origin"]
+        interp_order = kwargs["interp_order"]
+        domain_size = kwargs["domain_size"]
+        cell_width = kwargs["cell_width"]
+        refinement_ratio = kwargs.get("refinement_ratio", 1)
+        refinement_boxes = kwargs.get("refinement_boxes", {})
+
+    domain_box = boxm.Box(0, nbr_cells - 1)
+    domain_layout = GridLayout(domain_box, origin, cell_width, interp_order)
+
+    coarse_particles = Particles(box=domain_box)
+
+    # copy domain particles and put them in ghost cells
+    particle_ghost_nbr = domain_layout.particleGhostNbr(interp_order)
+    box_extend = particle_ghost_nbr - 1
```

Comment:
the reason for this line is not clear to me

Reply:
it's just a temporary variable. the important lines are 97 to 101, and comment L90.
comment says we're taking domain particles and put them in ghost cells.
Let's take an exemple of a domain of say, lower/upper=  0/32. At interp order 1, say, ghost box is (-1,33).
What this code does is filling cells -1 and 33.
cell -1 is filled by selecting particles in cell 32 (domain.box.upper - box_extend, with box_extend=ghostnbr-1)
cell 33 is filled by selecting particle in cell 0